modAionImpl/src/org/aion/zero/impl/AionHub.java

```diff
@@ -151,72 +168,81 @@ public AionHub() {
+                        p2pMgr,
+                        chainConfig.createBlockHeaderValidator(),
+                        cfg.getNet().getP2p().inSyncOnlyMode());
 
         registerCallback();
-        this.p2pMgr.run();
+        p2pMgr.run();
 
         ((AionPendingStateImpl) this.mempool).setP2pMgr(this.p2pMgr);
 
         this.pow = new AionPoW();
         this.pow.init(blockchain, mempool, eventMgr);
+    }
 
-        if (cfg.getReports().isHeapDumpEnabled()) {
-            new Thread(
-                new TaskDumpHeap(this.start, cfg.getReports().getHeapDumpInterval(), reportsFolder),
-                "dump-heap")
-                .start();
-        }
+    static AionHub createForTesting(
+            CfgAion _cfgAion, AionBlockchainImpl _blockchain, AionRepositoryImpl _repository) {
+        return new AionHub(_cfgAion, _blockchain, _repository, true);
+    }
+
+    private AionHub(
+            CfgAion _cfgAion,
+            AionBlockchainImpl _blockchain,
+            AionRepositoryImpl _repository,
+            boolean forTest) {
+        initializeHub(_cfgAion, _blockchain, _repository, forTest);
     }
 
     private void registerCallback() {
         List<Handler> cbs = new ArrayList<>();
-        cbs.add(new ReqStatusHandler(syncLOG, this.blockchain, this.p2pMgr,
-            cfg.getGenesis().getHash()));
-        cbs.add(new ResStatusHandler(syncLOG, this.p2pMgr, this.syncMgr));
-        cbs.add(new ReqBlocksHeadersHandler(
-            syncLOG, this.blockchain, this.p2pMgr, this.cfg.getNet().getP2p().inSyncOnlyMode()));
-        cbs.add(new ResBlocksHeadersHandler(syncLOG, this.syncMgr, this.p2pMgr));
-        cbs.add(new ReqBlocksBodiesHandler(
-            syncLOG, this.blockchain, this.p2pMgr, this.cfg.getNet().getP2p().inSyncOnlyMode()));
-        cbs.add(new ResBlocksBodiesHandler(syncLOG, this.syncMgr, this.p2pMgr));
-        cbs.add(new BroadcastTxHandler(
-            syncLOG, this.mempool, this.p2pMgr, this.cfg.getNet().getP2p().inSyncOnlyMode()));
-        cbs.add(new BroadcastNewBlockHandler(syncLOG, this.propHandler, this.p2pMgr));
+        cbs.add(new ReqStatusHandler(syncLOG, blockchain, p2pMgr, cfg.getGenesis().getHash()));
+        cbs.add(new ResStatusHandler(syncLOG, p2pMgr, syncMgr));
+        boolean inSyncOnlyMode = cfg.getNet().getP2p().inSyncOnlyMode();
+        cbs.add(new ReqBlocksHeadersHandler(syncLOG, blockchain, p2pMgr, inSyncOnlyMode));
+        cbs.add(new ResBlocksHeadersHandler(syncLOG, syncMgr, p2pMgr));
+        cbs.add(new ReqBlocksBodiesHandler(syncLOG, blockchain, p2pMgr, inSyncOnlyMode));
+        cbs.add(new ResBlocksBodiesHandler(syncLOG, syncMgr, p2pMgr));
+        cbs.add(new BroadcastTxHandler(syncLOG, mempool, p2pMgr, inSyncOnlyMode));
+        cbs.add(new BroadcastNewBlockHandler(syncLOG, propHandler, p2pMgr));
         this.p2pMgr.register(cbs);
     }
 
-    /**
-     */
-    private void loadEventMgr() {
+    private void loadEventMgr(boolean forTest) {
```

Comment:
I recommend naming the boolean "shouldStart" -- would be clearer.  And, if you want, mention in a JavaDoc that it's there for test purposes